src/core/namespace.c

```diff
@@ -2245,32 +2252,69 @@ int setup_namespace(const NamespaceParameters *p, char **error_path) {
         if (r < 0)
                 return r;
 
-        if (p->tmp_dir) {
-                bool ro = streq(p->tmp_dir, RUN_SYSTEMD_EMPTY);
+        if (p->private_tmp_as_tmpfs) {
+                _cleanup_free_ char *tmp_dir = NULL, *var_tmp_dir = NULL;
+                MountEntry *tmp_entry, *var_tmp_entry;
 
-                MountEntry *me = mount_list_extend(&ml);
-                if (!me)
+                tmp_dir = path_join(p->private_tmp_dir, "tmp");
+                var_tmp_dir = path_join(p->private_tmp_dir, "var/tmp");
+                if (!tmp_dir || !var_tmp_dir)
                         return log_oom_debug();
 
-                *me = (MountEntry) {
+                tmp_entry = mount_list_extend(&ml);
+                var_tmp_entry = mount_list_extend(&ml);
+                if (!tmp_entry || !var_tmp_entry)
+                        return log_oom_debug();
+
+                r = append_tmpfs_mounts(
+                                &ml,
+                                (TemporaryFileSystem []){{(char *) p->private_tmp_dir, (char *) "mode=01777" NESTED_TMPFS_LIMITS}},
+                                /* n= */ 1);
+                if (r < 0)
+                        return r;
+
+                *tmp_entry = (MountEntry) {
+                        .source_malloc = TAKE_PTR(tmp_dir),
                         .path_const = "/tmp",
-                        .mode = ro ? MOUNT_PRIVATE_TMP_READ_ONLY : MOUNT_PRIVATE_TMP,
-                        .source_const = p->tmp_dir,
+                        .mode = MOUNT_BIND,
+                        .source_dir_mode = 01777,
+                        .create_source_dir = true,
                 };
-        }
+                *var_tmp_entry = (MountEntry) {
+                        .source_malloc = TAKE_PTR(var_tmp_dir),
+                        .path_const = "/var/tmp",
+                        .mode = MOUNT_BIND,
+                        .source_dir_mode = 01777,
+                        .create_source_dir = true,
+                };
```

Comment:
so this works correctly only because MOUNT_BIND > MOUNT_TMPFS, i.e. that bind mounts are applied after tmpfs, after sorting. Would be good to encode this here, i.e.:

```c
assert_cc(MOUNT_BIND > MOUNT_TMPFS);
```

just to catch if somebody reorders the enum one day.